lang/machine.md

```diff
@@ -68,19 +73,45 @@ impl<M: Memory> Machine<M> {
             func: start_fn,
             locals: Map::new(),
             caller_return_info: CallerReturnInfo {
-                // the start function should never return.
+                // The start function should never return.
                 next_block: None,
                 ret_place: None,
             },
             next_block: start_fn.start,
             next_stmt: Int::ZERO,
         };
+        let mut mem = M::new();
+        let mut global_ptrs = Map::new();
+
+        // Allocate every global.
+        for (global_name, global) in prog.globals {
+            let size = Size::from_bytes(global.bytes.len()).unwrap();
+            let alloc = mem.allocate(size, global.align)?;
+            global_ptrs.insert(global_name, alloc);
+        }
+
+        // Fill the allocations.
+        for (global_name, global) in prog.globals {
+            let mut bytes = global.bytes.map(|b|
+                match b {
+                    Some(x) => AbstractByte::Init(x, None),
+                    None => AbstractByte::Uninit
+                }
+            );
+            for (i, relocation) in global.relocations {
+                let ptr = global_ptrs[relocation.name].offset::<M>(relocation.offset).unwrap();
+                let encoded_ptr = encode_ptr::<M>(ptr);
+                bytes.write_subslice_at_index(i.bytes(), encoded_ptr);
+            }
+            mem.store(global_ptrs[global_name], bytes, global.align)?;
+        }
```

Comment:
Nit: it'd make sense to do this before setting up the first stack frame, since arguably globals are "more global" than the stack.